graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/NodeMetadataSocialProofGenerator.java

```diff
@@ -0,0 +1,182 @@
+import com.twitter.graphjet.algorithms.RecommendationType;
+import com.twitter.graphjet.algorithms.TweetIDMask;
+import com.twitter.graphjet.bipartite.NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph;
+import com.twitter.graphjet.bipartite.NodeMetadataMultiSegmentIterator;
+import com.twitter.graphjet.hashing.IntArrayIterator;
+
+import it.unimi.dsi.fastutil.bytes.*;
+import it.unimi.dsi.fastutil.ints.*;
+import it.unimi.dsi.fastutil.longs.*;
+
+/**
+ * NodeMetadataSocialProofGenerator shares similar logic with
+ * {@link com.twitter.graphjet.algorithms.counting.TopSecondDegreeByCount}.
+ * In the request, clients specify a seed user set (left nodes) and a node metadata set
+ * (right node's metadata).
+ * NodeMetadataSocialProofGenerator finds the intersection between the seed users (left nodes)
+ * and the given node metadata id set. It accomplishes this by traversing each outgoing edge from
+ * the seed set and each right node's metadata from those edges.
+ * Only node metadatas with at least one social proof will be returned to clients.
+ */
+public abstract class NodeMetadataSocialProofGenerator implements
+  RecommendationAlgorithm<NodeMetadataSocialProofRequest, SocialProofResponse> {
+
+  private static final int MAX_EDGES_PER_NODE = 500;
+  private static final Byte2ObjectMap<Long2ObjectMap<LongSet>> EMPTY_SOCIALPROOF_MAP =
+    new Byte2ObjectArrayMap<>();
+
+  private NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph graph;
+  // NodeMetadata (Int) -> Engagement (Byte) -> User (Long) -> Tweets (LongSet)
+  private final Int2ObjectMap<Byte2ObjectMap<Long2ObjectMap<LongSet>>> socialProofs;
+  // NodeMetadata (Int) -> Sum of social proof edges (Double)
+  private final Int2DoubleMap socialProofWeights;
+  protected RecommendationType recommendationType;
+  protected IDMask idMask;
+
+  public NodeMetadataSocialProofGenerator(
+    NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph graph,
+    TweetIDMask mask,
+    RecommendationType recommendationType
+  ) {
+    this.graph = graph;
+    this.idMask = mask;
+    this.recommendationType = recommendationType;
+    // Variables socialProofs and socialProofWeights are re-used for each request.
+    // We chose to use an ArrayMap since, on average, we will request social proof for very few
+    // (<10) metadatas in a single request.
+    this.socialProofs = new Int2ObjectArrayMap<>();
+    this.socialProofWeights = new Int2DoubleArrayMap();
+  }
+
+  private void updateSocialProofWeight(int metadataId, double weight) {
+    // We sum the weights of incoming leftNodes as the weight of the rightNode.
+    socialProofWeights.put(
+      metadataId,
+      weight + socialProofWeights.get(metadataId)
+    );
+  }
+
+  private void addSocialProof(int metadataId, byte edgeType, long leftNode, long rightNode) {
+    if (!socialProofs.containsKey(metadataId)) {
+      // We chose to use an ArrayMap here since we will at most have 5 edge types that we request
```

Comment:
1. will at most => will have at most
2. better not to say "at most" a specific max value in the library code. 

We choose ArrayMap over OpenHashMap since the request will have at most a few edge types, usually less than five.